core/src/main/java/tech/pegasys/web3signer/core/Runner.java

```diff
@@ -224,22 +234,62 @@ private void registerHttpHostAllowListHandler(final OpenAPI3RouterFactory openAp
+    final OpenApiSpecsExtractor openApiSpecsExtractor =
+        new OpenApiSpecsExtractor.OpenApiSpecsExtractorBuilder()
+            .withFixRelativeRefPaths(false)
+            .build();
 
+    final Map<Path, String> swaggerUIWebRoot = loadSwaggerUIStaticContent(openApiSpecsExtractor);
+    LOG.debug("/swagger-ui paths: {}", swaggerUIWebRoot.keySet());
+
+    // serve /swagger-ui/* from static content map
     router
-        .route(HttpMethod.GET, SWAGGER_ENDPOINT + "/web3signer.yaml")
-        .produces(CONTENT_TYPE_YAML)
-        .handler(ResponseContentTypeHandler.create())
-        .handler(routingContext -> routingContext.response().end(openApiSpecYaml));
+        .route(HttpMethod.GET, SWAGGER_ENDPOINT + "/*")
+        .handler(ctx -> swaggerUIHandler(swaggerUIWebRoot, ctx));
 
+    // Vertx 3.x doesn't handle paths without trailing / (such as /swagger-ui) , so handle it
+    // directly. The following code may be removed once upgrade to Vertx 4.x
     router
-        .routeWithRegex(HttpMethod.GET, SWAGGER_ENDPOINT + "|" + SWAGGER_ENDPOINT + "/*")
-        .produces(CONTENT_TYPE_TEXT_HTML)
-        .handler(ResponseContentTypeHandler.create())
-        .handler(routingContext -> routingContext.response().end(indexHtml));
+        .route(HttpMethod.GET, SWAGGER_ENDPOINT)
+        .handler(ctx -> swaggerUIHandler(swaggerUIWebRoot, ctx));
+  }
+
+  private void swaggerUIHandler(
+      final Map<Path, String> swaggerUIContents, final RoutingContext ctx) {
+    final Path incomingPath = Path.of(ctx.request().path());
+    if (swaggerUIContents.containsKey(incomingPath)) {
+      ctx.response()
+          .putHeader("Content-Type", CONTENT_TYPE_TEXT_HTML)
+          .end(swaggerUIContents.get(incomingPath));
+    } else {
+      ctx.fail(404);
+    }
+  }
+
+  private Map<Path, String> loadSwaggerUIStaticContent(
+      final OpenApiSpecsExtractor openApiSpecsExtractor) {
+    final Map<Path, String> swaggerUIContentMap =
+        openApiSpecsExtractor.getDestinationSpecPaths().stream()
+            .map(
+                path -> {
+                  try {
+                    final Path relativePath =
+                        openApiSpecsExtractor.getDestinationDirectory().relativize(path);
+                    final Path swaggerUIPath = Path.of(SWAGGER_ENDPOINT).resolve(relativePath);
+                    final String content = Files.readString(path, UTF_8);
+                    return Map.entry(swaggerUIPath, content);
+                  } catch (final IOException e) {
+                    throw new UncheckedIOException(e);
+                  }
+                })
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    // map /swagger-ui to index.html content
+    final String index_html =
+        swaggerUIContentMap.get(Path.of(SWAGGER_ENDPOINT).resolve("index.html"));
+    swaggerUIContentMap.put(Path.of(SWAGGER_ENDPOINT), index_html);
+    return swaggerUIContentMap;
```

Comment:
This is probably large enough that we want to move this into it's own handler now